scico/linop/xray/_xray.py

```diff
@@ -193,3 +194,116 @@ def _calc_weights(x0, dx, nx, angle, y0):
+        r"""
+        Args:
+            im: Input image.
+            P: (num_angles, 2, 4) array of homogeneous projection matrices.
+            det_shape: Shape of detector.
+        """
+
+        x = jnp.mgrid[: im.shape[0], : im.shape[1], : im.shape[2]]
+        # (v, 2, 3) X (3, x0, x1, x2) + (v, 2) -> (v, 2, x0, x1, x2)
+        Px = (
+            jnp.tensordot(P[..., :3], x, axes=[2, 0])
+            + P[..., 3, np.newaxis, np.newaxis, np.newaxis]
+        )
+
+        # calculate weight on 4 intersecting pixels
+        w = 0.5  # assumed <= 1.0
+        left_edge = Px - w / 2
+        to_next = jnp.minimum(jnp.ceil(left_edge) - left_edge, w)
+        ul_ind = jnp.floor(left_edge).astype("int32")
+        ul_ind = jnp.where(ul_ind < 0, max(det_shape), ul_ind)  # otherwise negative values wrap
+
+        ul_weight = to_next[:, 0] * to_next[:, 1] * (1 / w**2)
+        ur_weight = (w - to_next[:, 0]) * to_next[:, 1] * (1 / w**2)
+        ll_weight = to_next[:, 0] * (w - to_next[:, 1]) * (1 / w**2)
+        lr_weight = (w - to_next[:, 0]) * (w - to_next[:, 1]) * (1 / w**2)
+
+        num_views = len(P)
+        proj = jnp.zeros((num_views,) + det_shape, dtype=im.dtype)
+        view_ind = jnp.expand_dims(jnp.arange(num_views), range(1, 4))
+        proj = proj.at[view_ind, ul_ind[:, 0], ul_ind[:, 1]].add(ul_weight * im, mode="drop")
+        proj = proj.at[view_ind, ul_ind[:, 0] + 1, ul_ind[:, 1]].add(ur_weight * im, mode="drop")
+        proj = proj.at[view_ind, ul_ind[:, 0], ul_ind[:, 1] + 1].add(ll_weight * im, mode="drop")
+        proj = proj.at[view_ind, ul_ind[:, 0] + 1, ul_ind[:, 1] + 1].add(
+            lr_weight * im, mode="drop"
+        )
+        return proj
+
+
+def P_to_vectors(in_shape: Shape, P: ArrayLike, det_shape: Shape) -> ArrayLike:
+    """
+    Convert SCICO projection matrix into ASTRA vectors.
+
+    For 3D arrays,
+    in Astra, the dimensions go (slices, rows, columns) and (z, y, x);
+    in SCICO, the dimensions go (x, y, z).
+
+    For 2D arrays,
+    in Astra, the dimensions go (rows, columns) and (y, x);
+    in SCICO, the dimensions go (x, y).
+
+    In Astra, the x-grid (recon) is centered on the origin and the y-grid (projection) can move.
+    In SCICO, the x-grid origin is x[0, 0, 0], the y-grid origin is y[0, 0].
+
+    See https://astra-toolbox.com/docs/geom3d.html#projection-geometries parallel3d_vec.
+    """
+    # ray is perpendicular to projection axes
+    ray = np.cross(P[:, 0, :3], P[:, 1, :3])
+    # detector center comes from lifting the center index to 3D
+    y_center = np.array(det_shape) / 2
+    x_center = np.einsum("...mn,n->...m", P[..., :3], np.array(in_shape) / 2) + P[..., 3]
+    d = np.einsum("...mn,...m->...n", P[..., :3], y_center - x_center)  # (V, 2, 3) x (V, 2)
+    u = -P[:, 1, :3]
+    v = -P[:, 0, :3]
+    vectors = np.concatenate((ray, d, u, v), axis=1)  # (v, 12)
+    return vectors
+
+
+def astra_to_scico(vol_geom, proj_geom):
```

Comment:
Naming not symmetric with "P_to_vectors", for which this is the inverse (no?).

Reply:
This function should have a corresponding test.

Reply:
convert_xray_geometry("astra_parallel3d_vec", "scico_projections", ...)